azavea/pwd-unitybar
# Add React UnityBar component + development infrastructure

## Overview

This PR turns the PWD Unity Bar prototype into a React component. I initially developed it against EPA STAR and moved some of that work here once it was far enough along. All of the necessary functionality for EPA STAR's implemented, and almost all of the remainder is here, too.

I'll put a todo list in the notes section below encompassing either follow up issues for this repo or things which remain to be done for this PR.

This PR https://github.com/azavea/epa-star/pull/172 is open in the EPA STAR repo, although it's sightly behind the work here.

Some images of the current configuration in that app are here:

https://github.com/azavea/epa-star/issues/132#issuecomment-326421420

Connects https://github.com/azavea/epa-star/issues/132

## Demo

![pwd-unitybar](https://user-images.githubusercontent.com/4165523/29988239-355e562c-8f3a-11e7-9628-bf271d7167dd.gif)

## Notes

*Development Environment*

First commit here creates a development infrastructure using webpack, webpack-dev-server, etc. I don't think there's anything too controversial here, although I did upgrade libraries to the newest non-breaking version and that includes ESLint & ESLint-airbnb.

One notable departure from our usual project structure is that this is not set up in a Docker container -- it uses a local Node/npm installation and webpack-dev-server etc.

I also set up a Demo.jsx file which can be used to try out the component props. These are enumerated in the UnityBar.jsx component's propTypes declaration, but it's probably worth pulling these out into the README for documentation, too.

*Implementation*

I tried to keep the dependencies as few as possible, which means we've only got react, react-dom, and react-onclickoutside plus the loaders etc required for transpiling JSX etc to JS. That means there's no Lodash or Redux in the project.

ReactDOM seems to be the largest library by far -- it's 218k vs 44k for React and 5.3k for onclickoutside. However, that library's only needed for the demo version since any other React app we put this in would have its own instance of ReactDOM.

Tried to keep this otherwise fairly boring: the UnityBar component is stateful and tracks when to open or close the various dropdowns; the rest of the components are functions and intended to be single files which pick apart the props to pass down to their child components.

I set up some standard constants including the various PWD app URLs and enums for things like themes, plus defined propTypes for things we imagine using in the future, like custom actions and custom menu options. Some of these remain to be implemented, and there are a few `throw new Error` statements in the code to indicate as-yet-unusable features.

*Distribution*

I haven't yet set up the distribution stuff for this, and we can either handle it in this PR or try to handle it very soon in a separate issue.

Roughly the ideal for this would be that we make distribution releases via GitHub here, and then our projects using this component pull in the package via their package.json file -- albeit from GitHub rather than via NPM.

A few things are required to make this happen:

- this project needs a webpack.dist.config.js file with instructions for creating a small bundle that's minified and, ideally, one file, since we're including all the Sass rules in the JS itself. If that's not viable, then we'll need a minified JS file and a minified CSS file as output to import into whatever app

- we need a release process in this repo which would roll through the git flow release steps to create a versioned release, then output the build in a format which other projects can bring in. Not sure how this typically works, but I *think* we've pulled in forked NPM dependencies from Azavea repos in the past, so we should investigate that.

I think the distribution stuff should happen in a separate issue; I think a good course here would be

1. get this PR +1'd then merged when viable
2. very very temporarily manually pull this component back into the EPA STAR app as part of https://github.com/azavea/epa-star/pull/172
3. in the next sprint, get this repo set up to have a version of the Unity Bar which can just be pulled from the repo and used in projects -- and replace the UnityBar in EPA STAR with a dependency on this repo.

## Testing
- get this branch, then `npm install` and `npm start`
- visit localhost:7777 to see the header bar in the browser. click through things and verify that they're generally working
- open the Demo.jsx component and adjust the configuration according to the proptypes enumerated here:

https://github.com/azavea/pwd-unitybar/pull/3/commits/a618b0026ab692fc02c72707dc0989cb67bf49cd#diff-e664379d39b3e5a4219b741502f0c5efR187

Not all of them are fully implemented, but you should encounter a thrown error when you try something this can't handle yet. You should also check that the props available for the UnityBar generally match what's described here: https://github.com/azavea/pwd-unitybar/blob/develop/prototype/Readme.md#react-component-parameters
